File: src/Security/PathGuard.php
<?php

declare(strict_types=1);

namespace ServerLens\Security;

final class PathGuard
{
    /** @var string[] resolved allowed directories */
    private array $allowedDirs = [];

    /** @var array<string, string> name => resolved path */
    private array $allowedPaths = [];

    /**
     * @param array<array{name: string, path: string}> $sources
     */
    public function registerSources(array $sources): void
    {
        foreach ($sources as $source) {
            $name = $source['name'];
            $path = $source['path'];
            $type = $source['type'] ?? 'file';

            $resolved = realpath($path);
            if ($resolved === false) {
                fwrite(STDERR, "[ServerLens] Warning: path not found for source '{$name}': {$path}\n");
                $this->allowedPaths[$name] = $path;
                continue;
            }

            $this->allowedPaths[$name] = $resolved;

            if ($type === 'directory' || is_dir($resolved)) {
                $this->allowedDirs[] = rtrim($resolved, '/') . '/';
            }
        }
    }

    public function getResolvedPath(string $name): ?string
    {
        return $this->allowedPaths[$name] ?? null;
    }

    public function isAllowed(string $path): bool
    {
        $resolved = realpath($path);
        if ($resolved === false) {
            return false;
        }

        if (in_array($resolved, $this->allowedPaths, true)) {
            return true;
        }

        foreach ($this->allowedDirs as $dir) {
            if (str_starts_with($resolved, $dir)) {
                return true;
            }
        }

        return false;
    }

    public function validateSource(string $name): ?string
    {
        $path = $this->getResolvedPath($name);
        if ($path === null) {
            return null;
        }

        $resolved = realpath($path);
        if ($resolved === false) {
            return null;
        }

        if ($resolved !== $path && !$this->isInAllowedDir($resolved)) {
            return null;
        }

        return $resolved;
    }

    private function isInAllowedDir(string $path): bool
    {
        foreach ($this->allowedDirs as $dir) {
            if (str_starts_with($path, $dir)) {
                return true;
            }
        }

        return false;
    }
}

File: serverlens/security/path_guard.py
from __future__ import annotations

import os
import sys
from typing import Any


class PathGuard:
    def __init__(self) -> None:
        self._allowed_dirs: list[str] = []
        self._allowed_paths: dict[str, str] = {}

    def register_sources(self, sources: list[dict[str, Any]]) -> None:
        for source in sources:
            name = source["name"]
            path = source["path"]
            src_type = source.get("type", "file")

            resolved = self._realpath(path)
            if resolved is None:
                print(
                    f"[ServerLens] Warning: path not found for source '{name}': {path}",
                    file=sys.stderr,
                )
                self._allowed_paths[name] = path
                continue

            self._allowed_paths[name] = resolved

            if src_type == "directory" or os.path.isdir(resolved):
                self._allowed_dirs.append(resolved.rstrip("/") + "/")

    def get_resolved_path(self, name: str) -> str | None:
        return self._allowed_paths.get(name)

    def is_allowed(self, path: str) -> bool:
        resolved = self._realpath(path)
        if resolved is None:
            return False

        if resolved in self._allowed_paths.values():
            return True

        return any(resolved.startswith(d) for d in self._allowed_dirs)

    @staticmethod
    def _realpath(path: str) -> str | None:
        try:
            rp = os.path.realpath(path)
            return rp if os.path.exists(rp) else None
        except OSError:
            return None
